### ISSUE TITLE
InsertCharPre like autocmd for the command line

### ISSUE THREAD

**Issue by altermo** [neovim/neovim#23813]
### Problem

Currently the only way to capture key presses within cmdline is with `vim.on_key`. The problem with this functionality is that it is to broad: it captures everything including `ctrl-v` and `ctrl-k` which there's no easy way to filter. In insert mode there's already an autocmd called `InsertCharPre` which does the job by only capturing the key presses which are logical to capture (no `ctrl-v` nor `ctrl-k`). I wish there was an autocmd which was similar to `InsertCharPre` but for command mode.

### Expected behavior

The behaviour would be simple: it works exactly as `InsertCharPre` but for the command mode.

**Comment by justinmk** [comment]
No major objection since it "completes" an existing precedent. But note also:

- `nvim_parse_cmd()` can be used to build very sophisticated cmdline mappings.
- custom cmdline completion was recently implemented: https://github.com/neovim/neovim/pull/24808

Can you give a specific use case?

**Comment by zeertzjq** [comment]
The commit message of #24808 is a bit misleading. I think it's actually just a bugfix.

**Comment by altermo** [comment]
One of the use cases that I thought of is being able to detect input of all chars in a match (like utf8 alpha chars).

**Comment by justinmk** [comment]
> One of the use cases that I thought of is being able to detect input of all chars in a match (like utf8 alpha chars).

For what purpose though?

**Comment by altermo** [comment]
`vim.regex:match_str`
An example of detecting UTF8 alpha char would be:
```lua
vim.api.nvim_create_autocmd('CmdlineCharPre',{callback=function ()
    if vim.regex([=[\v[[=a=][=b=][=c=][=d=][=e=][=f=][=g=][=h=][=i=][=j=][=k=][=l=][=m=][=n=][=o=][=p=][=q=][=r=][=s=][=t=][=u=][=v=][=w=][=x=][=y=][=z=]]*]=]):match_str(vim.v.char) then
        vim.schedule_wrap(vim.notify)('UTF8 alpha char detected')
    end
end})
```

**Comment by justinmk** [comment]
Er... that doesn't answer my question. What feature would you implement with this detection? I'm trying to understand how it would compare to a mapping that uses `nvim_parse_cmd()` to make decisions instead.

**Comment by altermo** [comment]
The features I would use this for are mappings which are a big/dynamic table of keys (like mapping all `'iskeyword'`);
And making it easier to work with mapping collisions.
Otherwise, it should be the same as a normal cmd-line mapping.

**Comment by lewis6991** [comment]
Why can't `CmdLineChanged` address this?

**Comment by altermo** [comment]
Cause it gets activated when ctrl-v is used or when completion is used.

**Comment by lewis6991** [comment]
Can you provide context on the motivation for this. What are you trying to implement?

**Comment by clason** [comment]
Yes, but we're asking what you want this _for_ -- what is the _end_ goal here?

**Comment by altermo** [comment]
I don't exactly understand the question.
But if I understand right:
The reason I want it is that I was developing a plugin that worked with Insert mode and I wanted to port that to Cmdline mode, and the plugin was using InsertCharPre.

**Comment by lewis6991** [comment]
What does the plugin do?

Just trying to avoid any XY situations.

**Comment by altermo** [comment]
The plugin itself does auto-pairing, while the specific feature that I was rewriting is auto-balance space between pairs when any alpha character is inserted.
<details><summary><b>examples</b></summary>

+ `( |)` > `a` > `( a| )`
+ `(  |)` > `a` > `(  a|  )`
</details>

**Comment by altermo** [comment]
I now have another use case: I created a plugin that allows one to use insert-mappings in the cmd-line.
My use case would be to be able to map all the keys (including built-in mappings) from cmd-line to insert-mode.
